pkg/stores/proxyalpha/proxy_store.go

```diff
@@ -0,0 +1,693 @@
+type WarningBuffer []types.Warning
+
+// HandleWarningHeader takes the components of a kubernetes warning header and stores them
+func (w *WarningBuffer) HandleWarningHeader(code int, agent string, text string) {
+	*w = append(*w, types.Warning{
+		Code:  code,
+		Agent: agent,
+		Text:  text,
+	})
+}
+
+// RelationshipNotifier is an interface for handling wrangler summary.Relationship events.
+type RelationshipNotifier interface {
+	OnInboundRelationshipChange(ctx context.Context, schema *types.APISchema, namespace string) <-chan *summary.Relationship
+}
+
+type Store struct {
+	clientGetter   ClientGetter
+	notifier       RelationshipNotifier
+	cacheFactory   CacheFactory
+	cfInitializer  CacheFactoryInitializer
+	namespaceCache Cache
+	lock           sync.Mutex
+	columnSetter   SchemaColumnSetter
+}
+
+type CacheFactoryInitializer func() (CacheFactory, error)
+
+type CacheFactory interface {
+	CacheFor(fields [][]string, client dynamic.ResourceInterface, gvk schema.GroupVersionKind, namespaced bool) (factory.Cache, error)
+	Reset() error
+}
+
+// NewProxyStore returns a Store implemented directly on top of kubernetes.
+func NewProxyStore(c SchemaColumnSetter, clientGetter ClientGetter, notifier RelationshipNotifier, factory CacheFactory) (*Store, error) {
+	store := &Store{
+		clientGetter: clientGetter,
+		notifier:     notifier,
+		columnSetter: c,
+	}
+
+	if factory == nil {
+		var err error
+		factory, err = defaultInitializeCacheFactory()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	store.cacheFactory = factory
+	if err := store.initializeNamespaceCache(); err != nil {
+		logrus.Infof("failed to warm up namespace informer for proxy store in steve, will try again on next ns request")
+	}
+	return store, nil
+}
+
+// Reset locks the store, resets the underlying cache factory, and warm the namespace cache.
+func (s *Store) Reset() error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	if err := s.cacheFactory.Reset(); err != nil {
+		return err
+	}
+
+	if err := s.initializeNamespaceCache(); err != nil {
+		return err
+	}
+	return nil
+}
+
+func defaultInitializeCacheFactory() (CacheFactory, error) {
+	informerFactory, err := factory.NewCacheFactory()
+	if err != nil {
+		return nil, err
+	}
+	return informerFactory, nil
+}
+
+// initializeNamespaceCache warms up the namespace cache as it is needed to process queries using options related to
+// namespaces and projects.
+func (s *Store) initializeNamespaceCache() error {
+	buffer := WarningBuffer{}
+	nsSchema := baseNSSchema
+
+	// make sure any relevant columns are set to the ns schema
+	if err := s.columnSetter.SetColumns(context.Background(), &nsSchema); err != nil {
+		return fmt.Errorf("failed to set columns for proxy stores namespace informer: %w", err)
+	}
+
+	// build table client
+	client, err := s.clientGetter.TableAdminClient(nil, &nsSchema, "", &buffer)
+	if err != nil {
+		return err
+	}
+
+	// get fields from schema's columns
+	fields := getFieldsFromSchema(&nsSchema)
```

Comment:
The schema columns end up like this in the DB:
- `metadata.fields[0]`
- `metadata.fields[1]`
- `metadata.fields[2]`
- etc

My understanding is that if columns are modified (eg: by updating the CRDs's printed column field), the index might change for these columns/fields. Could this create an issue in that case, where the UI looks for index 0 but it's now in index 1? In fact, the SQL cache might end up having a mix of old field + new field, right?

Reply:
That is what the schema handler is for, it resets the database on crd change.

Reply:
Oh right, I missed that somehow, thanks!